### PR TITLE
plugin/log: support multi nameScope

### PR DESCRIPTION
**Pull request by ckcd**
### 1. Why is this pull request needed and what does it do?
A large amount of logging will reduce performance, as the [plugin/log](https://coredns.io/plugins/log/) says:

> Note that for busy servers logging will incur a performance hit.

Fortunately `log` plugin can set the name that should be log:
```
log [NAME] [FORMAT] {
    class CLASSES...
}
```
So we can only log the domains that we concerned.

But, for now it can only set ONE name, and it's hard to make some workaround in `corefile` if we want to log more domains.

So this PR changes the `log` plugin to support multi name, and the `corefile` looks like:
```
log [NAMES...] {
    format FORMAT
    class CLASSES...
}
```

### 2. Which issues (if any) are related?
#2417 

### 3. Which documentation changes (if any) need to be made?
See `log/README.md`

### CODE REVIEW

File: plugin/log/README.md [hunk 1]
```diff
@@ -23,20 +23,21 @@ log
 Or if you want/need slightly more control:
 
 ~~~ txt
-log [NAME] [FORMAT]
```
Comment (miekg) [inline plugin/log/README.md:26]: If two parameters where specified we already assume the last one if the format - that should also make this change smaller (I guess).

Looking at the code and docs I see we also don't mention the use of `{common}` or `{combined}` as a shortcut. I'll send a PR for that.

File: plugin/log/log.go [hunk 2]
```diff
@@ -28,42 +28,43 @@ type Logger struct {
 func (l Logger) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
 	state := request.Request{W: w, Req: r}
 	for _, rule := range l.Rules {
-		if !plugin.Name(rule.NameScope).Matches(state.Name()) {
```
Comment (miekg) [inline plugin/log/log.go]: Would initiating multiple rules with the new names make sense here? We mean none of this code needs to change

Comment (ckcd) [reply plugin/log/log.go]: Hi,

I think it needs change, because in origin design, one `NAME` is associate with a `FORMAT`, so now in this PR, when initiating multiple rules with the new names, we have two choice:

1. just set one `FORMAT` for one`NAME` as origin, and don't need to change code here.
2. set multi `NAME` with one `FORMAT`, just as i do. and changes code here.

for method 1, That is a workaround of this situation, but it needs a format and a class for each name in `[]Rule`, which causes more memory. that's why i send this PR.

so i chose method 2, which needs to change code here.

thanks.

File: plugin/log/README.md (hunk 1 above)
Comment (ckcd) [reply plugin/log/README.md:26]: @miekg  Hi,

Thanks for your review,

I understand the opinion of make smaller change, 

but for now the format is `NAME FORMAT`, i don't know how to changes to `NAMES... FORMAT` in the code because it's hard to judge the last parameter that whether it's a name or a format. Unless  changes to `FORMAT NAMES...` which also changes bigger.

Also, for this plugin, `NAME` or `NAMES` is necessary but `FORMAT` and `class` is optional, so maybe set these two in `{}` would be better.

Comment (miekg) [reply plugin/log/README.md:26]: No, we just say the last one is `FORMAT`, this is happening now anyway in the code so there would be no change

File: plugin/log/log.go (hunk 2 above)
Comment (miekg) [reply plugin/log/log.go]: 1. is the better way

The map in the rule is a bit unfortunate (and causes most memory). Looks like an easy thing to remove/change.

**Comment by ckcd** [comment]
@miekg 

Hi,

I have change the PR with your two opinions:

1. make the `FORMAT` be the last one as origin;
2. use method 1 in our conversation, which means do NOT need to change the `serveDNS`;

Would you please review it again?

Thank you !!!

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/log: s..." ]
>@miekg
>
>Hi,
>
>I have change the PR with your two opinions:
>
>1. make the `FORMAT` be the last one as origin;
>2. use method 1 in our conversation, which means do NOT need to change the `serveDNS`;
>
>Would you please review it again?
>
>Thank you !!!

Thanks. I'll take a look.

File: plugin/log/setup.go [hunk 3]
```diff
@@ -34,6 +36,7 @@ func logParse(c *caddy.Controller) ([]Rule, error) {
 
 	for c.Next() {
 		args := c.RemainingArgs()
+		length := len(rules)
 
 		if len(args) == 0 {
```
Comment (miekg) [inline plugin/log/setup.go]: while we're touching this, let's convert this if-else-if-else to a switch as well, looks somewhat cleaner.

File: plugin/log/setup.go [hunk 4]
```diff
@@ -51,45 +54,57 @@ func logParse(c *caddy.Controller) ([]Rule, error) {
...
+					format = args[last]
+				}
 			}
 
-			rules = append(rules, Rule{
-				NameScope: dns.Fqdn(args[0]),
-				Format:    format,
-				Class:     make(map[response.Class]struct{}),
-			})
+			args = args[:last]
```
Comment (miekg) [inline plugin/log/setup.go]: can we not just chopped of the last element in the `strings.Contains` bit above? Then `args` is always correct, and we don't need to last even.

File: plugin/log/setup.go [hunk 5]
```diff
@@ -51,45 +54,57 @@ func logParse(c *caddy.Controller) ([]Rule, error) {
...
+				rules = append(rules, Rule{
+					NameScope: dns.Fqdn(str),
+					Format:    format,
+					Class:     make(map[response.Class]struct{}),
+				})
+			}
 		}
 
 		// Class refinements in an extra block.
+		classes := make(map[response.Class]struct{})
```
Comment (miekg) [inline plugin/log/setup.go:82]: it's unclear to me why this change is needed?

File: plugin/log/setup.go [hunk 6]
```diff
@@ -51,45 +54,57 @@ func logParse(c *caddy.Controller) ([]Rule, error) {
...
 			}
 		}
-		if len(rules[len(rules)-1].Class) == 0 {
-			rules[len(rules)-1].Class[response.All] = struct{}{}
+		if len(classes) == 0 {
+			classes[response.All] = struct{}{}
+		}
+
+		for i := len(rules) - 1; i >= length; i -= 1 {
+			rules[i].Class = classes
```
Comment (miekg) [inline plugin/log/setup.go:107]: the class map is now shared between all rules? Might make sense, but should probably be done separately?

**Comment by miekg** [comment]
Starting to look good.

File: plugin/log/setup.go (hunk 6 above)
Comment (ckcd) [reply plugin/log/setup.go:107]: Yes, the class map is shared, maybe we can add some optional to set class map for each nameScope, but it needs to change the format of this plugin also, e.g. 
```
log FORMAT {
NAME1 CLASSES...
NAME2 CLASSES...
...
}
```

File: plugin/log/setup.go (hunk 5 above)
Comment (ckcd) [reply plugin/log/setup.go:82]: Reason of this change is because each nameScope share class map as we talk in next comment.

File: plugin/log/setup.go (hunk 3 above)
Comment (ckcd) [reply plugin/log/setup.go]: Good idea~ doing it~

File: plugin/log/setup.go (hunk 4 above)
Comment (ckcd) [reply plugin/log/setup.go]: awesome !! i will take it~

**Comment by ckcd** [comment]
Sorry for the twice force push because i forget to rebase master in first one.

change log:
1. using `switch` rather than `if/else`;
2. delete `last` even and chopped `args` directly;

and for the case of: class map shared for multi nameScope, i have some ideas but it needs more changes and thinking, maybe we should do it in another part.


and i don't understand why test for other irrelevant plugin is failed, test of this plugin is OK.
the error in travis is:
```
--- FAIL: TestStop (0.07s)
    --- FAIL: TestStop/Test_0 (0.01s)
        proxy_test.go:66: Test 0, Expected healthchecks to hit test server, got none
```
and it seems NO any  relation with this PR.
@miekg does it make sense ?

File: plugin/log/setup.go (hunk 6 above)
Comment (miekg) [reply plugin/log/setup.go:107]: ah yes I see. Let's not make this change right now though.

File: plugin/log/setup.go (hunk 3 above)
Comment (ckcd) [reply plugin/log/setup.go]: done, please check.

File: plugin/log/setup.go (hunk 4 above)
Comment (ckcd) [reply plugin/log/setup.go]: Hi,

I have make the changes, please have a check.

Thank you !!!
